crates/dht/src/hash.rs

```diff
@@ -0,0 +1,440 @@
+//! partitions are updated in lockstep, which makes reasoning about the space-time partitioning
+//! easier.
+//!
+//! This module must be informed about ops that have been stored. There is no active process here
+//! that can look for newly stored ops. When a batch of ops is stored, the [PartitionedHashes] must
+//! be informed and will split the ops into the right space partition based on the location of the op.
+//! Ops are then pushed to the inner time partition for each space partition. It is the time
+//! partitions that are responsible for updating the combined hash values.
+//!
+//! That completes the high level information for this module, what follows is a more detailed
+//! explanation of the design.
+//!
+//! The bit-depth of the location roughly determines how evenly ops are distributed across the
+//! partitions. For example, an 8-bit location can only map op hashes to 256 possible locations,
+//! which would not use the top half of the partitions here. A 32-bit location retains enough
+//! information to use the full range of partitions and can be calculated inside a 32-bit integer.
+//! A 16-bit or 64-bit location would work too, and the choice between the options is not
+//! quantified here.
+//!
+//! The choice of 512 partitions is a tradeoff between the minimum amount of data that a node must
+//! store to participate in the network and the amount of data that must be sent over the network
+//! to discover what op data needs to be fetched. Every node will store between 0 and 512
+//! partitions. For nodes that are choosing to store data, the minimum they can store is 1
+//! partition. Or in other words, 1/512th of the total data. To put this in context, an application
+//! that has a complete data set of 5TB would require a minimum of 10GB of storage dedicated to
+//! that app to be able to participate in the network.
+//! It is important to note though, that each partition is managing a [PartitionedTime] structure
+//! which requires some memory and processing to maintain. Nodes that take on more partitions will
+//! have more work to do and will have to send more data during gossip rounds. That happens when
+//! there are changes to data that has made it into a time slice and become part of a combined
+//! hash. Generally, nodes will be able to just sync "what's new" but for catch-up, this is true.
+//! As a consequence, a new network that hasn't yet gained enough members to start reducing how
+//! many partitions are covered by each node will have a higher overhead for each node.
+//! It is the responsibility of the gossip module to work out how to be efficient about sending
+//! the minimum amount of data required to keep all nodes up to date. However, the worst case is
+//! that 512 combined hashes must be sent to find out which partition has a mismatch for a given
+//! time slice. Over time, there will be more time slices to check.
```

Comment:
One more question, this module is about partitioning hash space and arcs. This worst case is when there's a mismatch of the top hash of space partitions, I believe. If the top hash mismatches, we'll need to compare all 512 combined hashes of all the space partitions, to find out which of them is mismatching. Why is the time slice here?

Reply:
I could do with drawing some pictures to go with what I'm writing :) Maybe we can embed SVGs into docs. I'll have a look as part of the gossip work.

What I'm getting at here is thinking about space like a disc. Then for the time slices you draw rings on that disc starting with the start time at the center of the disc and with time moving outwards.

There are then several ways to select from that data. You could take a disc out of the disc. That would involve collecting across all space segments and all time slices from the start time up to some limit of time slices.

You could take out a ring which would involve picking a single time slice id and then collecting across all space segments.

The answer is yes though, once you're collecting in some way and you find a mismatch at the "top level" of that structure, you will be forced to send a whole lot more data. How much data you have to send will depend on how much was included in the collection that you found a mismatch for. I think that's going to be the crux of designing the diffing logic. How do we make it likely that we send smaller amounts of data more often than we have to send a lot.

Reply:
There's more work to be done on terminology and documentation here. I'm going to move forwards here and continue refining this writing as my ideas for how the diffing should work materialize